### PR TITLE
project loader, schema: add advanced grammar support for build-environment

### PR DESCRIPTION
**Pull request by cjp256**

### CODE REVIEW

**Comment by cjp256** [comment]
Grammar changes are complete, I'm going to split that into it's own PR, with build-environment changes coming separately.

**Comment by sergiusens** [comment]
Units are failing

**Review by sergiusens — APPROVED** [review]

File: snapcraft/internal/pluginhandler/__init__.py [hunk 1]
```diff
@@ -630,9 +628,6 @@ def _generate_part_env(self, step: steps.Step) -> str:
         else:
             plugin_environment = dict()
 
-        # Part's (user) say.
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py:633]: this code comment had a nice flow

File: snapcraft/internal/pluginhandler/__init__.py [hunk 2]
```diff
@@ -646,7 +641,7 @@ def _generate_part_env(self, step: steps.Step) -> str:
             for k, v in plugin_environment.items():
                 print(f'export {k}="{v}"', file=run_environment)
             print("## User Environment", file=run_environment)
-            for env in user_build_environment:
+            for env in self.build_environment:
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py:644]:
```suggestion
            # Part's (user) say.
            for env in self.build_environment:
```

Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py:644]: or (I prefer this even if it is more code) combined with the change some lines of code prior

```suggestion
            for env in user_build_environment:
```

File: snapcraft/internal/pluginhandler/__init__.py [hunk 3]
```diff
@@ -630,9 +628,6 @@ def _generate_part_env(self, step: steps.Step) -> str:
         else:
             plugin_environment = dict()
 
-        # Part's (user) say.
-        user_build_environment = self._part_properties["build-environment"]
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py:634]: code suggestions not working for removed lines, I would keep user_build_environment but = self.build_environment

This makes the code obvious to casual readers. The "user" part of this variable name was your idea ;-)

**Review by sergiusens — CHANGES_REQUESTED** [review]

File: snapcraft/internal/pluginhandler/__init__.py (hunk 3 above)
Comment (cjp256) [reply snapcraft/internal/pluginhandler/__init__.py:634]: Makes sense, will do

Comment (cjp256) [reply snapcraft/internal/pluginhandler/__init__.py:634]: Done.

File: snapcraft/internal/pluginhandler/__init__.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/internal/pluginhandler/__init__.py]: Done.

Comment (cjp256) [reply snapcraft/internal/pluginhandler/__init__.py]: Done.

**Comment by cjp256** [comment]
Only test failing is plainbox.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**